## Estadistica univariada
Los datos son parte del material del primer curso Ciencia de Datos de la Universidad de los Andes, por medio de la plataforma coursera. Aunque existe un método que nos puede brindar toda esta información, se tomó el tiempo de revisar de manera individual el como obtener cada una de esta información estadistica. 

### Objetivos 
1. Entender y analizar los uso de las tablas de frecuencia para resolver provlemas reelevantes para el negocio 
2. Entender y analizar las proncipales medidas de tendencia central y su interpretacion 
3. Entender y analizar las medidas de dispersion mas comunes para el analisis de datos

## 1. Contecto del caso
Se analizan las medidas de disperción, posición y variabilidad para el sector retail, para eso tomaremos los datos mas reelevantes de precios, medidas y envios de los productos para poder validar las hipotesis propuestas para cada tema.

## 2. Importacion de librerias y archivos

In [16]:
#libreria para comando de sistema
import os
#libreria para manejo de datos
import pandas as pd

In [17]:
#cargar los datos en csv
data = pd.read_csv('Ordenes_productos_C1_M2.csv', sep = ';', encoding = 'latin-1')

## 3.Lectura de datos
Es importante visualizar la forma en la que esta estructurada la tabla, asi como los tipos de datos que se manejan

In [18]:
data.head(5)

orden_id order_item_id producto_id vendedor_id fecha_envio_limite  \
0  107500PO59A             A        PO59      VE5389    7/04/2018 18:12   
1   37493PS22B             B        PS22      VE1558    20/10/2017 9:07   
2   28050PK20B             B        PK20      VE9159    17/08/2017 8:15   
3   52187PA10A             A        PA10      VE3159   23/09/2017 23:27   
4   84639PR12A             A        PR12      VE5090    7/01/2018 11:50   

   precio  valor_flete  codigo_postal_vendedor ciudad_vendedor  \
0  271.86        30.72                   52435         Mallama   
1  115.73         4.68                   52203           Colon   
2  432.99        82.70                   66001         Pereira   
3  108.38        35.39                   52435         Mallama   
4   51.50        11.10                   73001          Ibague   

  departamento_vendedor nombre_categoria_producto  longitud_nombre_producto  \
0                Nariño    Productos ecoamigables                       6.0   
1                Nariño                Carnicería                      10.0   
2             Risaralda                  Deportes                      25.0   
3                Nariño         Electrodomésticos                      10.0   
4                Tolima         Frutas y verduras                      23.0   

   longitud_descripcion_producto  cantidad_fotos_producto  peso_g_producto  \
0                            7.0                     27.0           2486.0   
1                           31.0                     20.0            256.0   
2                            5.0                      4.0           5270.0   
3                            1.0                      6.0            734.0   
4                           16.0                     35.0            884.0   

   longitud_cm_producto  altura_cm_producto  ancho_cm_producto  
0                  17.0                11.0               14.0  
1                  43.0                 2.0               21.0  
2                   9.0                27.0               29.0  
3                  46.0                48.0               22.0  
4                  45.0                26.0               18.0

In [19]:
data.dtypes

orden_id                          object
order_item_id                     object
producto_id                       object
vendedor_id                       object
fecha_envio_limite                object
precio                           float64
valor_flete                      float64
codigo_postal_vendedor             int64
ciudad_vendedor                   object
departamento_vendedor             object
nombre_categoria_producto         object
longitud_nombre_producto         float64
longitud_descripcion_producto    float64
cantidad_fotos_producto          float64
peso_g_producto                  float64
longitud_cm_producto             float64
altura_cm_producto               float64
ancho_cm_producto                float64
dtype: object

## 4. Tablas de frecuencia
Para la solucion del caso de uso es necesario identificar una posible hipótesis para validad a traves de las herramientas que se revisaran en las siguientes lineas.

La tabla de frecuencia es una heramienta que nos permite ordenar los datos, de manera que se presentan numericamente las caracteristicas de la distribucion de un conjunto de datos o muestra.

### Hipotesis:
La categoria de celulares es la que mas producto vende, se quiere validar ya que quiere lanzar una campaña de descuento para traer mas clientes a la empresa, sin embargo se desea conocer cual es la categoria en la que se invertirá mas publicidad.

In [20]:
##Tabla de frecuencia univariada
#esto para ver que categoria se vende mas
#Frecuencia absoluta

frec_data = data['nombre_categoria_producto'].value_counts().reset_index()
frec_data.head(4)

index  nombre_categoria_producto
0         Tecnología                        542
1           Deportes                        537
2  Frutas y verduras                        531
3         Carnicería                        523

In [23]:
#frecuencia relativa
frec_data['porcentaje_nombre_categoria_producto'] = frec_data['nombre_categoria_producto'].apply(lambda x: x/frec_data['nombre_categoria_producto'].sum())
frec_data.head(5)

index  nombre_categoria_producto  \
0         Tecnología                        542   
1           Deportes                        537   
2  Frutas y verduras                        531   
3         Carnicería                        523   
4    Ropa de adultos                        521   

   porcentaje_nombre_categoria_producto  
0                              0.053605  
1                              0.053110  
2                              0.052517  
3                              0.051726  
4                              0.051528

In [28]:
#la frecumencia realiva y absoluta acumulada usando cumsum()
frec_data['frec absoluta acumulada'] = frec_data['nombre_categoria_producto'].cumsum()
frec_data['frec relativa acumulada'] = frec_data['porcentaje_nombre_categoria_producto'].cumsum()

In [29]:
frec_data

index  nombre_categoria_producto  \
0               Tecnología                        542   
1                 Deportes                        537   
2        Frutas y verduras                        531   
3               Carnicería                        523   
4          Ropa de adultos                        521   
5               Ferretería                        518   
6                 Juguetes                        516   
7            Ropa infantil                        511   
8        Electrodomésticos                        509   
9               Dormitorio                        508   
10                 Lácteos                        499   
11  Productos ecoamigables                        497   
12                  Libros                        497   
13                Mascotas                        495   
14         Tarjetas regalo                        495   
15                   Salud                        488   
16               Celulares                        488   
17                 Muebles                        486   
18                  Bebés                         475   
19                   Licor                        475   

    porcentaje_nombre_categoria_producto  frec absoluta acumulada  \
0                               0.053605                      542   
1                               0.053110                     1079   
2                               0.052517                     1610   
3                               0.051726                     2133   
4                               0.051528                     2654   
5                               0.051231                     3172   
6                               0.051034                     3688   
7                               0.050539                     4199   
8                               0.050341                     4708   
9                               0.050242                     5216   
10                              0.049352                     5715   
11                              0.049154                     6212   
12                              0.049154                     6709   
13                              0.048957                     7204   
14                              0.048957                     7699   
15                              0.048264                     8187   
16                              0.048264                     8675   
17                              0.048066                     9161   
18                              0.046979                     9636   
19                              0.046979                    10111   

    frec relativa acumulada  
0                  0.053605  
1                  0.106715  
2                  0.159233  
3                  0.210958  
4                  0.262486  
5                  0.313718  
6                  0.364751  
7                  0.415290  
8                  0.465631  
9                  0.515874  
10                 0.565226  
11                 0.614380  
12                 0.663535  
13                 0.712491  
14                 0.761448  
15                 0.809712  
16                 0.857976  
17                 0.906043  
18                 0.953021  
19                 1.000000

In [30]:
## frecuencia bivariada
#tambla que nos muestra la frecuencia por ciudad de venta y categoria del producto
pd.crosstab(index = data['nombre_categoria_producto'], columns = data['ciudad_vendedor'])

ciudad_vendedor            Abejorral  Abrego  Abriaqui  Acacias  Acandi  \
nombre_categoria_producto                                                 
Bebés                              9      23         1       20       0   
Carnicería                         8      16         1       15       0   
Celulares                          7      19         0       17       0   
Deportes                          10      19         0       24       1   
Dormitorio                         8       9         2       19       0   
Electrodomésticos                  5      27         1       17       1   
Ferretería                         5      14         0       18       0   
Frutas y verduras                  6      20         0       16       2   
Juguetes                           9      18         0       19       2   
Libros                             7      18         1       20       2   
Licor                              6      16         0       14       0   
Lácteos                            6      12         1       19       0   
Mascotas                           9      12         0       13       1   
Muebles                            3      19         1       19       0   
Productos ecoamigables             8      15         0       20       1   
Ropa de adultos                    4      20         1       20       1   
Ropa infantil                     12      22         2       15       0   
Salud                              5      16         1       11       0   
Tarjetas regalo                    9      18         1       21       1   
Tecnología                         8      19         2       20       1   

ciudad_vendedor            Acevedo  Achi  Agua de dios  Arauca  Arboleda  ...  \
nombre_categoria_producto                                                 ...   
Bebés                            0     9             4       3         6  ...   
Carnicería                       0     5             7      10         7  ...   
Celulares                        1    11             9       4         9  ...   
Deportes                         0     4             3       5         5  ...   
Dormitorio                       0     9             3      11         9  ...   
Electrodomésticos                0     5             6       3         4  ...   
Ferretería                       0     6             7       9         6  ...   
Frutas y verduras                0     6             5       8        12  ...   
Juguetes                         0    11             3       6        11  ...   
Libros                           0     8             9       5         8  ...   
Licor                            1    10             4       9         2  ...   
Lácteos                          3     9             8       7         6  ...   
Mascotas                         0     7             7       4         7  ...   
Muebles                          2     5             3       1        11  ...   
Productos ecoamigables           2     9             4       5        13  ...   
Ropa de adultos                  0    11             3       3        10  ...   
Ropa infantil                    2    11             3      11        11  ...   
Salud                            0     6             4      10         3  ...   
Tarjetas regalo                  0     9             5      13         4  ...   
Tecnología                       1     6             4       3         8  ...   

ciudad_vendedor            San andres  San jose del guaviare  \
nombre_categoria_producto                                      
Bebés                               2                      6   
Carnicería                          1                      6   
Celulares                           3                      3   
Deportes                            3                      6   
Dormitorio                          2                      6   
Electrodomésticos                   2                      1   
Ferretería                          3                 

### Análisis:
De la tabla de frecuencias podemos observar que en frecuencia aucmulada los productos de tecnologia son los más vendidos

## 5. Tendencia central (moda, media y mediana)

### Hipótesis: 
los precios en el comercio son mayores a 500, ya que el estudio de marketing arrojó que los clientes más recurrentes compran mayormente productos más económicos (debajo de 500) por lo que es necesario analizar si en promedio la compañia tiene precios mas competitivos.

In [31]:
#media o poromedio de precios
data['precio'].mean()

119.7923465561476

In [32]:
#Mediana. Valor central de la distribucion de los datos
data['precio'].median()

83.12

In [33]:
#Moda. Los precios que mas se repiten 
data['precio'].mode()

0     3.92
1    11.39
2    11.79
3    15.80
4    16.42
5    25.45
6    34.43
7    36.96
8    56.79
dtype: float64

### Análisis: 
Es posible analizar el promedio y se observa que se encnuentra por debajos de los 500, localizandoce en los 120, por lo que en promedio se compran productos cuyo precio se encuentra por debajo de los 120 pesos, dentro de los cuales los productos con precio menor a 60 son lo smas comprados.

## 6. Medidas de localización
### Hipótesis:
la mitad (50%) de los productos pesa mas de 500g, ya que la empresa de mensajeria quiere realizar un descuento si los productos pesan menos de 500g, asi la compañia podria bajar los precios de envio de la mitad de los productos que se venden.

In [34]:
##Cuantiles
# calcular el 20%, 50% y 70% del peso de los productos 
data['peso_g_producto'].quantile([0.20, 0.50, 0.70])

0.2     471.0
0.5    1459.0
0.7    2536.0
Name: peso_g_producto, dtype: float64

### Análisis:
podemos ver que el cuantil 50% está por arriba de los 500g, por lo que mas de la mitad de los productos pesan mas de 500g, lo cual nos rechaza la hipótesis.

## 7. Medidas de disperción
### Hipótesis: 
La variacion de las longitudes de los datos es mayor a 20cm, ya que las de empresas de mensajeria que tienen cubiculos de transporte de mensajeria individual de producto de longitud de 20 cm son mas economicas, asi se podria cambiar a estas empresas y ahorrar en costos de envio.

In [35]:
##Desviacion estandar
#localizar la desviacion estandar de la ongitud de los productos de la mucestra
#representa que tan lejos de lamedias e encuentran las observaciones
data['longitud_cm_producto'].std()

15.286298866596708

In [36]:
##Rango intercuartilico
#encontrar el rango intercuartilico de la longitud de los productos
#es la diferencia entre el tercer y primer cuartil
data['longitud_cm_producto'].quantile(0.75) - data['longitud_cm_producto'].quantile(0.25)

22.0

In [37]:
##Rango 
#muestra la diferencia del producto mas grande con el mas chico
data['longitud_cm_producto'].max() - data['longitud_cm_producto'].min()

91.0

In [38]:
## Coeficiente de variacion 
# diviosn de la desciacion estander entre la sobre la media 
data['longitud_cm_producto'].std() / data['longitud_cm_producto'].mean()

0.5084018928267704

### Analisis: 
A pesar que la desviacion estandar y el rango intercuartilico son menos a 25, se puede analizar ocn el rando que la variacion entre el máximo y el mínimo es de 91cm, por lo tanto se podria migrar al rededor de 50% de los preductos de las empresas de mensajeria con el cubivulo de 20cm y darle al cliente un descuento en el envio.

### Método describe()
Este método nos permite obtener toda la inforamcion anterior de una manera rápida. A continuación se usa el método y como se puede observar coincide con la información obtenida anteriormente.

In [39]:
data.describe()

precio   valor_flete  codigo_postal_vendedor  \
count  10134.000000  10134.000000            10134.000000   
mean     119.792347     20.053246            49991.684626   
std      119.340785     20.106163            22463.895340   
min        0.010000      0.000000             5001.000000   
25%       34.155000      5.760000            50001.000000   
50%       83.120000     13.980000            52323.000000   
75%      165.515000     27.455000            54003.000000   
max     1262.940000    183.150000            99001.000000   

       longitud_nombre_producto  longitud_descripcion_producto  \
count              10111.000000                   10111.000000   
mean                  20.070023                      19.844724   
std                   11.719844                      11.851631   
min                    0.000000                       0.000000   
25%                   10.000000                      10.000000   
50%                   20.000000                      20.000000   
75%                   30.000000                      30.000000   
max                   40.000000                      40.000000   

       cantidad_fotos_producto  peso_g_producto  longitud_cm_producto  \
count             10111.000000     10111.000000          10111.000000   
mean                 20.097616      2106.475522             30.067352   
std                  11.802373      2094.402206             15.286299   
min                   0.000000         0.000000              0.000000   
25%                  10.000000       607.000000             19.000000   
50%                  20.000000      1459.000000             30.000000   
75%                  30.000000      2941.000000             41.000000   
max                  40.000000     20661.000000             91.000000   

       altura_cm_producto  ancho_cm_producto  
count        10111.000000        10111.00000  
mean            17.911186           22.91257  
std             11.359838           11.15189  
min              0.000000            0.00000  
25%              9.000000           15.00000  
50%             17.000000           23.00000  
75%             26.000000           30.00000  
max             64.000000           72.00000